<h1>Tabla de Contenidos<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Cross-Validation-para-Selección-de-Modelo-e-Hiperparámetros:---&quot;Separando-el-Test-Set&quot;" data-toc-modified-id="Cross-Validation-para-Selección-de-Modelo-e-Hiperparámetros:---&quot;Separando-el-Test-Set&quot;-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Cross Validation para Selección de Modelo e Hiperparámetros:   "Separando el Test Set"</a></span></li><li><span><a href="#GridSearchCV" data-toc-modified-id="GridSearchCV-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>GridSearchCV</a></span></li><li><span><a href="#Ejemplo-1-:-Cross-Validation-para-Selección-de-Modelo-e-Hiperparámetros:---&quot;Separando-el-Test-Set&quot;-usando-GridSearchCV-(más-&quot;automático&quot;)." data-toc-modified-id="Ejemplo-1-:-Cross-Validation-para-Selección-de-Modelo-e-Hiperparámetros:---&quot;Separando-el-Test-Set&quot;-usando-GridSearchCV-(más-&quot;automático&quot;).-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Ejemplo 1 : Cross Validation para Selección de Modelo e Hiperparámetros:   "Separando el Test Set" usando GridSearchCV (más "automático").</a></span><ul class="toc-item"><li><span><a href="#Cargamos-los-datos-y-separamos-en-Train-y-Test-en-proporción-80/20-y-con-semilla-123" data-toc-modified-id="Cargamos-los-datos-y-separamos-en-Train-y-Test-en-proporción-80/20-y-con-semilla-123-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Cargamos los datos y separamos en Train y Test en proporción 80/20 y con semilla 123</a></span></li><li><span><a href="#Modelo-de-Árbol:" data-toc-modified-id="Modelo-de-Árbol:-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Modelo de Árbol:</a></span><ul class="toc-item"><li><span><a href="#Hiperparámetros" data-toc-modified-id="Hiperparámetros-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>Hiperparámetros</a></span></li><li><span><a href="#Creamos-el-GridSearchCV" data-toc-modified-id="Creamos-el-GridSearchCV-3.2.2"><span class="toc-item-num">3.2.2&nbsp;&nbsp;</span>Creamos el GridSearchCV</a></span></li><li><span><a href="#.fit---Entrenamiento--del-GridSearchCV" data-toc-modified-id=".fit---Entrenamiento--del-GridSearchCV-3.2.3"><span class="toc-item-num">3.2.3&nbsp;&nbsp;</span>.fit   Entrenamiento  del GridSearchCV</a></span></li><li><span><a href="#Veamos-los-resultados" data-toc-modified-id="Veamos-los-resultados-3.2.4"><span class="toc-item-num">3.2.4&nbsp;&nbsp;</span>Veamos los resultados</a></span><ul class="toc-item"><li><span><a href="#.best_score_" data-toc-modified-id=".best_score_-3.2.4.1"><span class="toc-item-num">3.2.4.1&nbsp;&nbsp;</span>.best_score_</a></span></li><li><span><a href="#.best_params_" data-toc-modified-id=".best_params_-3.2.4.2"><span class="toc-item-num">3.2.4.2&nbsp;&nbsp;</span>.best_params_</a></span></li><li><span><a href="#.best_estimator_" data-toc-modified-id=".best_estimator_-3.2.4.3"><span class="toc-item-num">3.2.4.3&nbsp;&nbsp;</span>.best_estimator_</a></span></li><li><span><a href="#.cv_results_" data-toc-modified-id=".cv_results_-3.2.4.4"><span class="toc-item-num">3.2.4.4&nbsp;&nbsp;</span>.cv_results_</a></span></li><li><span><a href="#Accuracy-promedio-para-cada-valor-del-hiperparámetro:" data-toc-modified-id="Accuracy-promedio-para-cada-valor-del-hiperparámetro:-3.2.4.5"><span class="toc-item-num">3.2.4.5&nbsp;&nbsp;</span>Accuracy promedio para cada valor del hiperparámetro:</a></span></li><li><span><a href="#Desvío-Standard-de-Accuracy-para-cada-valor-del-hiperparámetro:" data-toc-modified-id="Desvío-Standard-de-Accuracy-para-cada-valor-del-hiperparámetro:-3.2.4.6"><span class="toc-item-num">3.2.4.6&nbsp;&nbsp;</span>Desvío Standard de Accuracy para cada valor del hiperparámetro:</a></span></li><li><span><a href="#.best_index_" data-toc-modified-id=".best_index_-3.2.4.7"><span class="toc-item-num">3.2.4.7&nbsp;&nbsp;</span>.best_index_</a></span></li><li><span><a href="#Desvío-Standard-de-Accuracy-del-mejor-modelo" data-toc-modified-id="Desvío-Standard-de-Accuracy-del-mejor-modelo-3.2.4.8"><span class="toc-item-num">3.2.4.8&nbsp;&nbsp;</span>Desvío Standard de Accuracy del mejor modelo</a></span></li><li><span><a href="#AC-+/--2-*--desvio--(95%)" data-toc-modified-id="AC-+/--2-*--desvio--(95%)-3.2.4.9"><span class="toc-item-num">3.2.4.9&nbsp;&nbsp;</span>AC +/- 2 *  desvio  (95%)</a></span></li></ul></li><li><span><a href="#Resultados-de-GridSearchCV-como-DataFrame" data-toc-modified-id="Resultados-de-GridSearchCV-como-DataFrame-3.2.5"><span class="toc-item-num">3.2.5&nbsp;&nbsp;</span>Resultados de GridSearchCV como DataFrame</a></span></li></ul></li><li><span><a href="#Modelo-de-Regresión-Logística-+-Ridge------Autoevaluación--" data-toc-modified-id="Modelo-de-Regresión-Logística-+-Ridge------Autoevaluación---3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Modelo de Regresión Logística + Ridge    - Autoevaluación -</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#.best_score_" data-toc-modified-id=".best_score_-3.3.0.1"><span class="toc-item-num">3.3.0.1&nbsp;&nbsp;</span>.best_score_</a></span></li><li><span><a href="#.best_params_" data-toc-modified-id=".best_params_-3.3.0.2"><span class="toc-item-num">3.3.0.2&nbsp;&nbsp;</span>.best_params_</a></span></li><li><span><a href="#.best_estimator_" data-toc-modified-id=".best_estimator_-3.3.0.3"><span class="toc-item-num">3.3.0.3&nbsp;&nbsp;</span>.best_estimator_</a></span></li><li><span><a href="#.cv_results_" data-toc-modified-id=".cv_results_-3.3.0.4"><span class="toc-item-num">3.3.0.4&nbsp;&nbsp;</span>.cv_results_</a></span></li><li><span><a href="#Accuracy-promedio-para-cada-valor-del-hiperparámetro:" data-toc-modified-id="Accuracy-promedio-para-cada-valor-del-hiperparámetro:-3.3.0.5"><span class="toc-item-num">3.3.0.5&nbsp;&nbsp;</span>Accuracy promedio para cada valor del hiperparámetro:</a></span></li><li><span><a href="#Desvío-Standard-de-Accuracy-del-mejor-modelo" data-toc-modified-id="Desvío-Standard-de-Accuracy-del-mejor-modelo-3.3.0.6"><span class="toc-item-num">3.3.0.6&nbsp;&nbsp;</span>Desvío Standard de Accuracy del mejor modelo</a></span></li><li><span><a href="#AC-+/--2-*--desvio--(95%)" data-toc-modified-id="AC-+/--2-*--desvio--(95%)-3.3.0.7"><span class="toc-item-num">3.3.0.7&nbsp;&nbsp;</span>AC +/- 2 *  desvio  (95%)</a></span></li></ul></li></ul></li></ul></li><li><span><a href="#Selección-de-Modelo-y-Evaluación-del-modelo-elegido-en-el-Test-Set-con-.score" data-toc-modified-id="Selección-de-Modelo-y-Evaluación-del-modelo-elegido-en-el-Test-Set-con-.score-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Selección de Modelo y Evaluación del modelo elegido en el Test Set con .score</a></span><ul class="toc-item"><li><span><a href="#Selección-de-Modelo" data-toc-modified-id="Selección-de-Modelo-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Selección de Modelo</a></span></li><li><span><a href="#Evaluación-de-Accuracy-frente-al-Test-Set-&quot;Separado&quot;" data-toc-modified-id="Evaluación-de-Accuracy-frente-al-Test-Set-&quot;Separado&quot;-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Evaluación de Accuracy frente al Test Set "Separado"</a></span><ul class="toc-item"><li><span><a href="#.score" data-toc-modified-id=".score-4.2.1"><span class="toc-item-num">4.2.1&nbsp;&nbsp;</span>.score</a></span></li></ul></li></ul></li><li><span><a href="#Modelo-para-Producción" data-toc-modified-id="Modelo-para-Producción-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Modelo para Producción</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#.fit-para-entrenar" data-toc-modified-id=".fit-para-entrenar-5.0.1"><span class="toc-item-num">5.0.1&nbsp;&nbsp;</span>.fit para entrenar</a></span></li><li><span><a href="#.predict-para-pronosticar" data-toc-modified-id=".predict-para-pronosticar-5.0.2"><span class="toc-item-num">5.0.2&nbsp;&nbsp;</span>.predict para pronosticar</a></span></li></ul></li><li><span><a href="#Guardado-para-uso-posterior" data-toc-modified-id="Guardado-para-uso-posterior-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Guardado para uso posterior</a></span><ul class="toc-item"><li><span><a href="#Uso-posterior-del-modelo:" data-toc-modified-id="Uso-posterior-del-modelo:-5.1.1"><span class="toc-item-num">5.1.1&nbsp;&nbsp;</span>Uso posterior del modelo:</a></span></li></ul></li></ul></li></ul></div>

# Cross Validation para Selección de Modelo e Hiperparámetros

## Cross Validation para Selección de Modelo e Hiperparámetros:   "Separando el Test Set"

Sci-Kit Learn nos provee con muchas herramientas que pueden ayudarnos mucho en nuestra tarea, por ejemplo contamos con GridSearchCV que nos permitirá hacer más fácil todavía el procedimiento anterior.   

Primero veamos qué hace GridSearchCV y luego veamos cómo utilizarlo en nuestro procedimiento de Cross Validation para Selección de Modelo e Hiperparámetros "Separando el Test Set" usando el mismo ejemplo que en el caso anterior.

Recordemos el flujo de trabajo que vamos a realizar

<img src="https://i.ibb.co/Ky0Q9HC/grid-search-workflow.png" alt="grid-search-workflow" border="0" width=400>

## GridSearchCV   

> GridSearchCV nos permitirá seleccionar **para un modelo dado** y usando Cross Validation el **mejor valor para su/s hiperparámetro / s** , en forma casi "automática", **nos evitará tener que hacer un for** para iterar por cada valor posible de cada hiperparámetro.  

**Nota:** Habrá notado que hemos usado el plural para hiperparámetro / s, esto es así porque puede ser que algún modelo tenga más de uno, lo cual obviamente complica su búsqueda. Hasta ahora hemos hecho ejemplos con uno solo por fines didácticos, pero a partir de ahora podremos resolver este problema fácilmente. 

Supongamos que queremos resolver el mismo problema anterior: determinar para los modelos de Árbol y Regresión Logística con Ridge,  cuáles son los mejores valores para sus respectivos hiperparámetro / s y además queremos hacerlo con Cross Validation para evaluar con más confianza los resultados, en definitiva lo que hacíamos en el tema anterior con un **for** pero ahora vamos a hacerlo con GridSearchCV.

La documentación oficial: https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html

La sintaxis es la siguiente
~~~
sklearn.model_selection.GridSearchCV(estimator, param_grid, *, scoring=None, n_jobs=None, iid='deprecated', refit=True, cv=None, verbose=0, pre_dispatch='2*n_jobs', error_score=nan, return_train_score=False)
~~~
Los parámetros principales son: 

- **estimator**: lo mismo que antes, es el modelo que queremos utilizar
- **param_grid**: Es un **diccionario** con el nombre de cada hiperparámetro y los valores que queremos que pruebe. Por ejemplo si el modelo que estamos utilizando admite 3 hiperparámetros llamados a, b y c, sería:   
~~~
{'a':[1,2,3,4], 'b':[1,10,100,1000], 'c':[0.0001,0.001,0.01]}  
~~~
donde hemos indicado para cada parámetro los valores que queremos que pruebe. **Los nombres de los hiperparámetros deben ser los que admite el modelo en cuestión** no los podemos inventar.

- **scoring:** la métrica de evaluación que deseamos utilizar, por defecto el valor es None lo cual significa que utilizará el score por defecto que tenga el modelo, que para el caso de Clasificación suele ser AC y para el de Regresión suele ser $R^2$. Pero si lo deseamos podemos indicarle otro, o incluso **varios**. 

- **refit**: Por defecto en True, nos permitirá al final calcular la métrica del modelo evaluado en todo el dataset que le hayamos pasado. Dejar en True.

- **cv:** cantidad de folds, por defecto 5 
- **return_train_score=False** False por defecto, pero puede ser de interés obtener también los valores de score correspondientes al train set.  

- **n_jobs=None** Si nuestro procesador posee más de un núcleo, podemos indicar cuántos debe usar, si el valor es -1 usa todos los que tiene.

Vamos a ver cómo lo podemos utilizar!

## Ejemplo 1 : Cross Validation para Selección de Modelo e Hiperparámetros:   "Separando el Test Set" usando GridSearchCV (más "automático").

### Cargamos los datos y separamos en Train y Test en proporción 80/20 y con semilla 123

In [ ]:
import pandas as pd
df=pd.read_csv('data/datos1.csv')


X=df.drop('y', axis='columns')
y=df['y']

Ahora separamos en Train y Test:

In [ ]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=20,random_state=123)

In [ ]:
from sklearn.model_selection import GridSearchCV

### Modelo de Árbol:

Creamos el modelo de árbol

In [ ]:
from sklearn.tree import DecisionTreeClassifier
arbol=DecisionTreeClassifier(random_state=123)  # para que sea repetible

#### Hiperparámetros

**Armamos el Diccionario con los nombres y valores de los Hiperparámetros**:  

Aqui es donde podríamos pasar más de un hiperparámetro ... y nos evitaremos hacer un for dentro de otro for, etc si tenemos varios.   
En este caso sólo estamos interesados en la profuncidad, **max_depth** (este nombre debe ser el que corresponde al modelo, por las dudas, fijarse en la documentación oficial):  
https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html )


In [ ]:
parametros_arbol = {'max_depth':[1,2,3,4]}   

#### Creamos el GridSearchCV

In [ ]:
grid_arbol = GridSearchCV(estimator = arbol,param_grid = parametros_arbol,scoring = 'accuracy', cv = 5, verbose = 1,n_jobs = -1)

# Usé verbose=1 para que nos cuente un poco qué hace, puede omitirse.


#### .fit   Entrenamiento  del GridSearchCV
Lo "entrenamos" con .fit en nuestros datos.    

**Observe que le debemos pasar el Train Set para que entrene, porque queremos que cree los folds dentro de él sin tocar el Test Set que dejamos aparte. Al dividir al Train Set, cada una de las partes corresponderá a lo que llamamos ValTrain y ValTest.** que es lo que queríamos hacer.

In [ ]:
grid_arbol.fit(X_train, y_train);   # el ";" al final es para que no muestre toodos los parámetros que va a usar.

Fitting 5 folds for each of 4 candidates, totalling 20 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done  20 out of  20 | elapsed:    2.0s finished


**Listo** Ya está todo ... **sin un for** para iterar en los valores del hiperparámetro (o varios si hubieran sido varios hiperparámetros!). Para cada valor de hiperparámetro (o combinación de ellos si eran varios) hizo los folds, entrenó en lo que llamamos ValTrain y evaluó Accuracy en el ValTest de cada fold, también determinó qué parámetro arrojó mejor resultado (Accuracy en este caso) y varias cosas más!

#### Veamos los resultados

##### .best_score_  
**Cuál fue el mejor resultado obtenido para la métrica de evaluación (en este caso Accuracy)?**

In [ ]:
AC_arbol_best=grid_arbol.best_score_
AC_arbol_best

0.8333333333333334

##### .best_params_ 
**Cuál fue el valor de hiperparámetro/ s con el cual se obtuvo el mejor score anterior?**

**Nota:** Si hubiéramos pasado más hiperparámetros en el diccionario aqui nos hubiera dado la combinación de ellos que mejor resultado obtuvo!

In [ ]:
profundidad_arbol_best=grid_arbol.best_params_
profundidad_arbol_best

{'max_depth': 2}

##### .best_estimator_

**Cuál fue el mejor modelo?**  

Aquí nos devolvuelve todo el modelo ya **entrenado y con el mejor valor de Hiperparámetro/s** en este caso profundidad 2 y el resto de los parámetros fijos que le pasamos nosotros. **Este modelo incluye todo lo necesario para usarse luego para pronosticar, etc. **

In [ ]:
modelo_arbol_best=grid_arbol.best_estimator_
modelo_arbol_best

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=2, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=123, splitter='best')

##### .cv_results_  
**Queremos ver un resumen de todo lo que hizo?**

In [ ]:
grid_arbol.cv_results_

{'mean_fit_time': array([0.00358968, 0.00658231, 0.00418859, 0.00678205]),
 'std_fit_time': array([0.00048788, 0.00522451, 0.00097747, 0.00473747]),
 'mean_score_time': array([0.00159593, 0.00219407, 0.00219474, 0.00179534]),
 'std_score_time': array([0.00048829, 0.00193385, 0.00039918, 0.00039926]),
 'param_max_depth': masked_array(data=[1, 2, 3, 4],
              mask=[False, False, False, False],
        fill_value='?',
             dtype=object),
 'params': [{'max_depth': 1},
  {'max_depth': 2},
  {'max_depth': 3},
  {'max_depth': 4}],
 'split0_test_score': array([0.83333333, 0.83333333, 0.66666667, 0.66666667]),
 'split1_test_score': array([0.66666667, 0.83333333, 0.83333333, 0.83333333]),
 'split2_test_score': array([0.66666667, 0.66666667, 0.5       , 0.5       ]),
 'split3_test_score': array([0.83333333, 1.        , 1.        , 1.        ]),
 'split4_test_score': array([0.83333333, 0.83333333, 0.83333333, 0.83333333]),
 'mean_test_score': array([0.76666667, 0.83333333, 0.766666

Observe que presenta los resultados con un diccionario, así que si queremos acceder a sólo alguno de estos valores, podemos hacerlo de la siguiente manera, seguramente estaremos interesados en dos de estos resultados, el promedio de Accuracy ('mean_test_score') y el Desvío Standard 'std_test_score', tenga en cuenta que los resultados se presentan en el orden de los valores del hiperparámetro:

In [ ]:
grid_arbol.cv_results_['params']

[{'max_depth': 1}, {'max_depth': 2}, {'max_depth': 3}, {'max_depth': 4}]

Es decir que el primer resultado corresponde a profundidad 1, el siguiente a profundidad 2, etc.

##### Accuracy promedio para cada valor del hiperparámetro:

In [ ]:
grid_arbol.cv_results_['mean_test_score']

array([0.76666667, 0.83333333, 0.76666667, 0.76666667])

Cómo se lee? 

- Para el el primer valor de hiperparámetro, es decir profundidad 1, el promedio de Accuracy en los 5 folds fue de 0.766,
- Para el segundo, profundidad 2 el valor promedio de Accuracy fue de 0.8333, y así con el resto.
Puede observar que efectivamente cuando la profundidad es 2 obtuvo una Accuracy de 0.8333 que es la mejor de todas.

##### Desvío Standard de Accuracy para cada valor del hiperparámetro:

In [ ]:
desvios_arbol=grid_arbol.cv_results_['std_test_score']
desvios_arbol

array([0.08164966, 0.10540926, 0.16996732, 0.16996732])

Se lee de la misma manera que el anterior:

Podemos ver que el desvío standard para profundidad 2 es de 0.1054

Leyendo los valores podemos escribir Promedio +/- 2Desvío Sandard  para la profundidad 2 que es la que dio mejor AC como hicimos anteriormente:

El mejor Accuracy ya lo teníamos ( en AC_arbol_best), pero el Desvío deberíamos escribirlo **manualmente** (oh, oh!):

In [ ]:
print("Mejor Accuracy: ",AC_arbol_best,' +/- ', 2*0.10540926, '( 95% )')

Mejor Accuracy:  0.8333333333333334  +/-  0.21081852 ( 95% )


##### .best_index_
**Acceso programático al desvío correspondiente al mejor Accuracy**

Algunas personas se quejan porque dicen que hay que leer literalmente el valor del desvío para el mejor caso, sin embargo **también es posible acceder programáticamente** a cualquier valor de desvío (o de media, etc, cualquier cosa que esté en cv_results_) si uno quiere automatizar todo.   

Ocurre que GridSearchCV también tiene el atributo **best_index_**, que nos permite acceder programáticamente al valor del  desvío correspondiente al hiperparámetro que anduvo mejor!  

Fijémonos:

In [ ]:
grid_arbol.best_index_

1

Nos devolvió 1 porque si bien era el segundo elemento (profundidad 2), corresponde a índice 1 porque Python numera desde 0.  

Entonces al array  que llamamos 
~~~
desvios_arbol = grid_arbol.cv_results_['std_test_score'] 
~~~
simplemente deberíamos pasarle el índice 
~~~
[grid_arbol.best_index_] 
~~~

##### Desvío Standard de Accuracy del mejor modelo

In [ ]:
desvio_arbol_mejor=desvios_arbol[grid_arbol.best_index_]
desvio_arbol_mejor

0.10540925533894599

##### AC +/- 2 *  desvio  (95%)

In [ ]:
print("Modelo de Árbol")
print("Mejor Accuracy: ",AC_arbol_best,' +/- ', 2*desvio_arbol_mejor, '( 95% )')

Modelo de Árbol
Mejor Accuracy:  0.8333333333333334  +/-  0.21081851067789198 ( 95% )


Y lo pudimos hacer todo programáticamente! Por supuesto no es necesario hacerlo así, tranquilamente podemos observar los resultados y tomar nota a mano. 

#### Resultados de GridSearchCV como DataFrame

Algo más arriba usamos 
~~~
grid_arbol.cv_results_
~~~

y vimos que nos presentaba todos los resultados posibles en formato de **diccionario**, si prefiere ... hasta lo puede pasar a un DataFrame!

In [ ]:
resultados_arbol=pd.DataFrame(grid_arbol.cv_results_)
resultados_arbol

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0       0.003590      0.000488         0.001596        0.000488   
1       0.006582      0.005225         0.002194        0.001934   
2       0.004189      0.000977         0.002195        0.000399   
3       0.006782      0.004737         0.001795        0.000399   

  param_max_depth            params  split0_test_score  split1_test_score  \
0               1  {'max_depth': 1}           0.833333           0.666667   
1               2  {'max_depth': 2}           0.833333           0.833333   
2               3  {'max_depth': 3}           0.666667           0.833333   
3               4  {'max_depth': 4}           0.666667           0.833333   

   split2_test_score  split3_test_score  split4_test_score  mean_test_score  \
0           0.666667           0.833333           0.833333         0.766667   
1           0.666667           1.000000           0.833333         0.833333   
2           0.500000           1.000000           0.833333         0.766667   
3           0.500000           1.000000           0.833333         0.766667   

   std_test_score  rank_test_score  
0        0.081650                2  
1        0.105409                1  
2        0.169967                2  
3        0.169967                2

y podemos acceder a cualquier elemento en la tradicional forma provista por Pandas!  

Por supuesto que podemos seguir usando el best_index_  en el DataFrame, como por ejemplo si queremos sólo uno de sus valores, por ejemplo el valor del hiperparámetro para el mejor caso:

In [ ]:
resultados_arbol.loc[grid_arbol.best_index_,'params']

{'max_depth': 2}

O  bien podemos obtener todos los resultados correspondientes al mejor modelo!

In [ ]:
# Resúmen para el Mejor Árbol:
resultados_arbol.loc[grid_arbol.best_index_,:]

mean_fit_time              0.00658231
std_fit_time               0.00522451
mean_score_time            0.00219407
std_score_time             0.00193385
param_max_depth                     2
params               {'max_depth': 2}
split0_test_score            0.833333
split1_test_score            0.833333
split2_test_score            0.666667
split3_test_score                   1
split4_test_score            0.833333
mean_test_score              0.833333
std_test_score               0.105409
rank_test_score                     1
Name: 1, dtype: object

También puede resultar de mucho interés la última columna del DataFrame que nos ofrece un ranking para cada caso...

### Modelo de Regresión Logística + Ridge    - Autoevaluación - 


Se anima a hacer Ud solo el otro modelo que habíamos usado en el procedimiento anterior pero ahora usando GridSearchCV?  

- Era Regresión Logísitca con Regularización Ridge
- Los valores para el hiperparámetro C eran [1e-6,1e-5,1e-4,0.001,0.01,0.1,1,10]
- al GridSearchCV llámelo **grid_RL**

Obtenga lo mismo que antes:  

- Mejor valor de Accuracy +/- 2 desvíos
- Valor del Hiperparámetro C para el cual se obtuvo.
- Mejor modelo de RL + Ridge

In [ ]:
from sklearn.linear_model import LogisticRegression

# Creamos el modelo con los parámetros que no cambiarán, observe que no le pasamos el valor de C de regulrización
Log_Ridge=LogisticRegression(penalty='l2', max_iter=10000, tol=0.0001) 

# Armamos el diccionario con el  nombre y valores para los Hiperparámetros
parametros_RL = {'C':[1e-6,1e-5,1e-4,0.001,0.01,0.1,1,10]}   

# Armamos el GridSearchCV
grid_RL = GridSearchCV(estimator = Log_Ridge,param_grid = parametros_RL,scoring = 'accuracy', cv = 5, verbose = 1,
                        n_jobs = -1)

# Lo entrenamos en el Train Set (él lo dividirá internamente y en cada fold estarán en ValTrain y ValTest)
grid_RL.fit(X_train, y_train);
# Listo!

Fitting 5 folds for each of 8 candidates, totalling 40 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done  40 out of  40 | elapsed:    0.0s finished


##### .best_score_

In [ ]:
# Obtenemos el mejor AC 
AC_RL_best=grid_RL.best_score_
AC_RL_best

0.9

In [ ]:
# Que se obtuvo para un valor del hiperparámetro C de:

##### .best_params_

In [ ]:
C_RL_best=grid_RL.best_params_
C_RL_best

{'C': 0.1}

##### .best_estimator_

**Recordemos que aquí tenemos el modelo completo con todo lo necesario inclusive para pronosticar, etc**

In [ ]:
modelo_RL_best=grid_RL.best_estimator_
modelo_RL_best

LogisticRegression(C=0.1, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=10000,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

##### .cv_results_

Ahora que ya sebemos qué es, no veo necesidad de mostrarlo, así que paso.

##### Accuracy promedio para cada valor del hiperparámetro:

No lo necesitamos

In [ ]:
# grid_RL.cv_results_['mean_test_score']

##### Desvío Standard de Accuracy del mejor modelo

In [ ]:
# Podemos calcular el desvío standard para el mejor caso:
desvios_RL = grid_RL.cv_results_['std_test_score'] 
desvio_mejor_RL=desvios_RL[grid_RL.best_index_]
desvio_mejor_RL

0.08164965809277258

##### AC +/- 2 *  desvio  (95%)

In [ ]:
print("Modelo de Regresión Logística  con Regularización Ridge")
print("Mejor Accuracy: ",AC_RL_best,' +/- ', 2*desvio_mejor_RL, '( 95% )')

Modelo de Regresión Logística  con Regularización Ridge
Mejor Accuracy:  0.9  +/-  0.16329931618554516 ( 95% )


## Selección de Modelo y Evaluación del modelo elegido en el Test Set con .score

### Selección de Modelo

Comparemos los resultados obtenidos para el Árbol y RL + Ridge: 

In [ ]:
print("Modelo de Árbol")
print("Mejor Accuracy: ",AC_arbol_best,' +/- ', 2*desvio_arbol_mejor, '( 95% )')

print("------------------------------------------------------------------")
print("Modelo de Regresión Logística  con Regularización Ridge")
print("Mejor Accuracy: ",AC_RL_best,' +/- ', 2*desvio_mejor_RL, '( 95% )')

Modelo de Árbol
Mejor Accuracy:  0.8333333333333334  +/-  0.21081851067789198 ( 95% )
------------------------------------------------------------------
Modelo de Regresión Logística  con Regularización Ridge
Mejor Accuracy:  0.9  +/-  0.16329931618554516 ( 95% )


Nos decidiríamos por el Regresión Logística con Ridge.

### Evaluación de Accuracy frente al Test Set "Separado"

Una vez elegido el mejor modelo, ahora tenemos que **entrenarlo en todo el Train Set** para luego **evaluarlo en el Test Set** que mantuvimos separado a tal efecto, para tener una idea del valor de Accuracy que obtendremos frente a las observaciones no vistas, por suerte **GridSearchCV también nos ayuda para ello!**   

#### .score

grid_RL fue el ganador, entonces simplemente haciendo: 

~~~
grid_RL.score(X_test, y_test)
~~~

**"ya sabe" que tiene que entrenar en todo el TrainSet, usando el mejor hiperparámetro (C=0.1 en este caso)** ... muy bueno! Sólo pasándole (X_test, y_test) sabe que tienen que aplicar el modelo entrenado a las observaciones del X_test, calcular los pronósticos para cada una y luego calcular la Accuracy frente a los valores conocidos dle y_test!

**Hace todo solo!**


In [ ]:
AC=grid_RL.score(X_test, y_test)
AC

1.0

Nota: si el modelo elegido hubiera sido el árbol el código hubiera sido:

~~~
grid_arbol.score(X_test, y_test)
~~~

>**Sólo podemos lamentar que no presentamos los valores de AC con sus correspondientes intervalos de confianza.** porque hemos evaluado sólo en un Test Set.

## Modelo para Producción

Dónde tenemos guardados todos los parámetros correspondientes al mejor modelo de RL?  

-  en **grid_RL.best_estimator_** 

llamémoslo modelo_produccion para que sea más claro:

In [ ]:
modelo_produccion=grid_RL.best_estimator_

Este modelo_produccion se puede entrenar con .fit y se lo puede usar para pronosticar con .predict ... como cualquier modelo que hemos utilizado!

>**El modelo para producción debe ser entrenado en todo el DataSet original** ya no se evalúa porque asumimos el valor anterior como el válido.

#### .fit para entrenar 

en todo el DataSet

In [ ]:
modelo_produccion.fit(X,y)

LogisticRegression(C=0.1, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=10000,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

#### .predict para pronosticar  

supongamos que queremos pronosticar la y para una nueva observación:  x1=1, x2=1.5 x3=0.87

In [ ]:
modelo_produccion.predict([[1,1.5,0.87]])

array([1], dtype=int64)

O también podemos hacerlo para todo un conjunto de ellas, como las que podrían venir en un DataFrame:

In [ ]:
df_nuevo=pd.DataFrame([[-0.5,0.5,2],[2,0,-3.1],[-1,1,1]], columns=['x1','x1','x3'])
df_nuevo

x1   x1   x3
0 -0.5  0.5  2.0
1  2.0  0.0 -3.1
2 -1.0  1.0  1.0

In [ ]:
modelo_produccion.predict(df_nuevo)

array([0, 1, 0], dtype=int64)

### Guardado para uso posterior

In [ ]:
import pickle
# Comenté las siguientes líneas porque sino cada vez que corro este archivo, guardaría.

archivo = 'data/modelo_produccion2.sav'
pickle.dump(modelo_produccion, open(archivo, 'wb'))

#### Uso posterior del modelo:

In [ ]:
archivo = 'data/modelo_produccion2.sav'
modelo_cargado=pickle.load(open(archivo, 'rb'))

# Bibliografía

https://scikit-learn.org/stable/modules/grid_search.html#grid-search